src/ByteCountingStream.php

```diff
@@ -0,0 +1,61 @@
+<?php
+namespace GuzzleHttp\Psr7;
+
+use Psr\Http\Message\StreamInterface;
+
+/**
+ *  Stream decorator that ensures an expected number of bytes can be read
+ *  from an underlying read stream. \RuntimeException is thrown when
+ *  the underlying stream fails to provide the expected number of bytes.
+ */
+class ByteCountingStream implements StreamInterface
+{
+    use StreamDecoratorTrait;
+
+    /** @var int Number of bytes remains to be read */
+    private $remaining;
+
+    /**
+     * @param StreamInterface $stream       Stream to wrap
+     * @param int             $bytesToRead  Number of bytes to read
+     * @throws \InvalidArgumentException
+     */
+    public function __construct(StreamInterface $stream, $bytesToRead)
+    {
+        $this->stream = $stream;
+
+        if (!is_int($bytesToRead) || $bytesToRead < 0) {
+            $msg = "Bytes to read should be non-negative integer, got {$bytesToRead}.";
+            throw new \InvalidArgumentException($msg);
+        }
+
+        if ($bytesToRead > $this->stream->getSize()) {
+            $msg = "The ByteCountingStream decorator expects to be able to "
+                . "read {$bytesToRead} from a stream, but the stream being decorated "
+                . "only contains {$this->stream->getSize()} bytes.";
+            throw new \InvalidArgumentException($msg);
+        }
+
+        $this->remaining = $bytesToRead;
+    }
+
+    public function read($length)
+    {
+        if ($this->remaining === 0) {
+            return '';
+        }
+
+        if ($length <= $this->remaining) {
```

Comment:
Add something like `$bytesToRead = min($length, $this->remaining);` and then attempt to read that number of bytes from the underlying stream. It's fine if the underlying stream cannot return the requested number of bytes, simply return whatever it gives you. If, however, it returns `''` and you haven't read the expected number of bytes, then that's a failure.